File: test-suite/src/test/java/io/micronaut/mqtt/docs/publisher/qos/ProductClient.java
package io.micronaut.mqtt.docs.publisher.qos;

// tag::imports[]
import io.micronaut.mqtt.annotation.Qos;
import io.micronaut.mqtt.annotation.Topic;
import io.micronaut.mqtt.v5.annotation.MqttPublisher;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "PublisherQosSpec")
// tag::clazz[]
@MqttPublisher
public interface ProductClient {

    @Topic(value = "product", qos = 2) // <1>
    void send(byte[] data);

    @Topic("product")
    void send(byte[] data, @Qos int qos); // <2>
}
// end::clazz[]

File: test-suite-python/src/test/python/micronaut/mqtt/docs/publisher/qos/ProductClient.py
from abc import ABC, abstractmethod
from typing import Annotated

# tag::imports[]
from micronaut.mqtt.annotation import Qos, Topic
from micronaut.mqtt.annotation.v5 import MqttPublisher
# end::imports[]

from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="PublisherQosSpec")
# tag::clazz[]
@MqttPublisher
class ProductClient(ABC):

    @Topic(value="product", qos=2)  # <1>
    @abstractmethod
    def send(self, data: bytes) -> None:
        ...

    @Topic("product")
    @abstractmethod
    def send_with_qos(self, data: bytes, qos: Annotated[int, Qos]) -> None:  # <2>
        ...
# end::clazz[]
